### ISSUE TITLE
ternary operator on the same array poisons gradients

### ISSUE THREAD

**Issue by wi-re** [NVIDIA/warp#1736]
### Bug Description

Hi,

There's an issue with warp when computing gradients over ternary operations. The bug manifests in the following pattern of access
```py
def access(flag: wp.bool, index: wp.int32, arr: wp.array(dtype=wp.float64), term: wp.array(dtype=wp.float64)):
   return arr[index] if flag else return arr[index]/term[index]
```
This is a seemingly reasonable statement to write and occurs in many cases for optional correction terms. The forward pass will compute the correct values, but the backward pass will silently poison the gradient and return all zeros. The bug occurs in warp 1.12 and warp 1.16 (diagnostics attached below).  Using a manual if/else block results in correct forwards and backward behavior and is a simple enough workaround for now (the reproduction script shows this working fine as well)

Here's a script that will reproduce the bug:
```py
import warp as wp

wp.init()


@wp.kernel
def ternary_kernel(
    arr: wp.array(dtype=wp.float64),
    term: wp.array(dtype=wp.float64),
    flag: wp.bool,
    out: wp.array(dtype=wp.float64),
):
    i = wp.tid()
    val = arr[i] / term[i] if flag else arr[i]  # both branches read arr[i] -- the dangerous shape
    out[i] = val


@wp.kernel
def ifelse_kernel(
    arr: wp.array(dtype=wp.float64),
    term: wp.array(dtype=wp.float64),
    flag: wp.bool,
    out: wp.array(dtype=wp.float64),
):
    i = wp.tid()
    if flag:
        val = arr[i] / term[i]
    else:
        val = arr[i]
    out[i] = val


def run(kernel, flag: bool, n: int = 4):
    device = "cpu"
    arr = wp.array(list(range(1, n + 1)), dtype=wp.float64, requires_grad=True, device=device)
    term = wp.array([2.0] * n, dtype=wp.float64, requires_grad=True, device=device)
    out = wp.zeros(n, dtype=wp.float64, requires_grad=True, device=device)

    tape = wp.Tape()
    with tape:
        wp.launch(kernel, dim=n, inputs=[arr, term, flag], outputs=[out], device=device)

    # Seed every output adjoint with 1.0, i.e. compute d(sum(out))/d(arr).
    seed = wp.array([1.0] * n, dtype=wp.float64, device=device)
    tape.backward(grads={out: seed})

    forward = out.numpy().copy()
    grad_arr = arr.grad.numpy().copy()
    tape.zero()
    return forward, grad_arr


n = 4
expected_forward = [float(v) for v in range(1, n + 1)]  # flag=False -> val = arr[i]
expected_grad = [1.0] * n  # d(arr[i])/d(arr[i]) = 1

print(f"{'kernel':<16}{'forward matches arr':<22}{'grad(arr) == 1.0 everywhere'}")
print("-" * 70)

for name, kernel in (("ternary", ternary_kernel), ("if/else", ifelse_kernel)):
    forward, grad_arr = run(kernel, flag=False, n=n)
    forward_ok = list(forward) == expected_forward
    grad_ok = list(grad_arr) == expected_grad
    print(f"{name:<16}{str(forward_ok):<22}{str(grad_ok)} (grad(arr) = {list(grad_arr)})")
```

The expected output of the reproduction script (ran today on google colab):
```
kernel          forward matches arr   grad(arr) == 1.0 everywhere
----------------------------------------------------------------------
ternary         True                  False (grad(arr) = [np.float64(0.0), np.float64(0.0), np.float64(0.0), np.float64(0.0)])
if/else         True                  True (grad(arr) = [np.float64(1.0), np.float64(1.0), np.float64(1.0), np.float64(1.0)]) 
```

### System Information

Software
  Warp (package):   1.16.0
  Warp (core):      1.16.0
  warp-clang:       1.16.0 (LLVM 18.1.3)
  NumPy:            2.0.2
  PyTorch:          2.11.0+cpu
  JAX:              0.7.2
  jaxlib:           0.7.2
  Python:           3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0]
  Platform:         Linux-6.6.122+-x86_64-with-glibc2.35

CUDA
  Enabled:          True
  Toolkit:          12.9
  NVRTC:            12.9
  Forward compat:   True

Libraries
  MathDx:           0.3.1
  cuBQL:            enabled
  NanoVDB:          32.8.0

Build
  Compiler:         GCC 14.2.1
  Debug:            False
  Verify FP:        False
  Fast math:        False
  Sanitizer:        none

CPU
  Model:            znver2
  ISA features:     39
                      64bit, adx, aes, avx, avx2, bmi, bmi2, clflushopt, ...

Devices
  cpu
    Name:             x86_64

{'warp_python': '1.16.0',
 'warp_native': '1.16.0',
 'warp_clang': '1.16.0 (LLVM 18.1.3)',
 'llvm': '18.1.3',
 'numpy': '2.0.2',
 'torch': '2.11.0+cpu',
 'jax': '0.7.2',
 'jaxlib': '0.7.2',
 'python': '3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0]',
 'platform': 'Linux-6.6.122+-x86_64-with-glibc2.35',
 'cuda_enabled': True,
 'cuda_toolkit': '12.9',
 'cuda_driver': None,
 'nvrtc': '12.9',
 'cuda_compatibility': True,
 'mathdx_enabled': True,
 'libmathdx': '0.3.1',
 'cubql_enabled': True,
 'nanovdb': '32.8.0',
 'host_compiler': 'GCC 14.2.1',
 'cpu_name': 'znver2',
 'cpu_features': ['64bit',
  'adx',
  'aes',
  'avx',
  'avx2',
  'bmi',
  'bmi2',
  'clflushopt',
  'clwb',
  'clzero',
  'cmov',
  'crc32',
  'cx16',
  'cx8',
  'f16c',
  'fma',
  'fsgsbase',
  'fxsr',
  'lzcnt',
  'mmx',
  'movbe',
  'pclmul',
  'popcnt',
  'prfchw',
  'rdpid',
  'rdrnd',
  'rdseed',
  'sahf',
  'sha',
  'sse',
  'sse2',
  'sse3',
  'sse4.1',
  'sse4.2',
  'sse4a',
  'ssse3',
  'xsave',
  'xsavec',
  'xsaveopt'],
 'debug': False,
 'verify_fp': False,
 'fast_math': False,
 'sanitizer': '',
 'devices': [{'alias': 'cpu', 'name': 'x86_64'}]}

**Comment by daedalus5** [comment]
Thanks for the thorough repro. A fix is underway.

**Comment by shi-eric** [comment]
Progress update: 8e9f0a97a296beec2544d76ecef019facfc9a42e landed a fix for ternary expressions whose branches read array elements but failed to propagate gradients.

The code generator now copies reference-valued branch results before merging them, which restores gradients for the selected array elements. We'll leave this issue open until @wi-re has a chance to try the fix in their workload.

### Public API behavior

- `arr[i] / term[i] if flag else arr[i]` and similar ternary expressions in `@wp.kernel` or `@wp.func` now propagate gradients through the selected branch. Their forward behavior is unchanged.
- This does not add a Python API or kernel builtin.

### Test coverage

- `warp/tests/test_conditional.py::test_ifexp_reference_branch_grad` covers both outcomes for scalar array values, including the original division case and selection between different arrays.
- `warp/tests/test_conditional.py::test_ifexp_reference_vec_branch_grad` covers both outcomes for vector array values.

### Behavioral verification

A separate CPU `float64` check followed the original repro for both flag values. The forward values and input gradients matched the expected results and the explicit `if/else` version.

**Comment by wi-re** [comment]
Hey @shi-eric,

Just checked out warp and built it locally, then ran my tests. This fixes the issue for the toy problem and fixes it in the actual codebase as well, thanks!

I did have some issues building warp in my environment, but that's a separate issue and probably down to how conda installed CUDA.

**Comment by shi-eric** [comment]
Sounds good, @daedalus5 was the one who made the fix. Let us know if we can help with the CUDA issue. There is also https://github.com/conda-forge/warp-lang-feedstock/issues if you narrow it down to the Conda recipe.
